### ISSUE TITLE
mAP is too low and val obj_loss is increasing while training is decreasing

### ISSUE THREAD

**Issue by sarmientoj24** [ultralytics/yolov5#7384]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have a YOLOv5 [run](https://wandb.ai/jamesandrewsarmiento/PAVE/runs/2ykdopsn?workspace=user-jamesandrewsarmiento) that involves small number of annotated dataset.

The mAP is too small, about 0.08 and the val obj_loss is increasing while the training is not. 
What could be the potential reason why obj_loss is increasing for validation? Anything that stands out as wrong from the config?

![image](https://user-images.githubusercontent.com/8830319/162811964-8ce8dd74-5f25-4267-8892-3dc4f9f03d79.png)

### Configs
```
# hyp-low
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 0.05
cls: 0.5
cls_pw: 1.0
obj: 1.0
obj_pw: 1.0
iou_t: 0.20
anchor_t: 4.0
# anchors: 3
fl_gamma: 0.0
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.0
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
```

```
# Modified augmentations
            T = [
                A.ImageCompression(quality_lower=75, p=0.2),
                A.RandomBrightnessContrast(
                    brightness_limit=0.1, contrast_limit=0.1, p=0.3
                ),
                A.OneOf(
                    [A.Blur(blur_limit=3, p=1.0), A.MedianBlur(blur_limit=3, p=1.0)], p=0.25
                ),
                # A.GaussNoise(var_limit=(0.0001, 0.0005), per_channel=False, p=0.1),
                A.OneOf([A.Sharpen(), A.Emboss()], p=0.25),
                # A.ToGray(p=0.1),
                A.CLAHE(p=0.25),
                A.ShiftScaleRotate(
                    shift_limit=0.0625, scale_limit=0.2, rotate_limit=25, p=0.15,
                ),
                A.OneOf(
                    [
                        A.RandomSizedBBoxSafeCrop(width=1280, height=768, p=1.0),
                        A.RandomSizedCrop([320, 720], 768, 1280, p=1.0)
                    ], p=0.15
                ), 
                A.Resize(width=1280, height=768, always_apply=True, p=1)
            ]  # transforms
```

Command
```
train.py --hyp data/hyps/hyp.scratch-low.yaml --data data.yaml --img 1280 --rect --epochs 200 --weights VisDrone/VisDrone2/weights/best.pt --project PAVE --name BATCH1_EXP3_Vis_PT --batch-size 16 --cache --workers 8 --device 1
```

Sample bbox debuger
![image](https://user-images.githubusercontent.com/8830319/162812711-6af3c578-39a5-43ab-b532-03374e0ae93e.png)

I actually also tried validating using the train images in the validation set just to check and it loooks quite good.
```
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 59/59 [00:35<00:00,  1.64it/s]          
                 all       1859       7131      0.922      0.817      0.894      0.598
                 D01       1859        445      0.966      0.905      0.951      0.605
                 D03       1859        368      0.951      0.842      0.945       0.64
                 D04       1859       1115      0.928      0.868      0.933      0.686
                 D06       1859       3252      0.972        0.9      0.964      0.663
                 D11       1859        655      0.883      0.656      0.751      0.468
                 D12       1859       1063      0.901      0.821      0.876      0.569
                 D13       1859        233      0.853      0.725      0.838      0.557
```

**Comment by glenn-jocher** [comment]
@sarmientoj24 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by sarmientoj24** [comment]
`labels.jpg`
![image](https://user-images.githubusercontent.com/8830319/162952663-36e0f7c9-95a6-4b17-8f46-3ed822395113.png)

`P_curve`
![image](https://user-images.githubusercontent.com/8830319/162952764-492ee958-21c4-4400-90f2-e66d1ff8c80f.png)

`Losses`
![image](https://user-images.githubusercontent.com/8830319/162952809-d0d3e819-53c8-409f-817f-890e37eb2888.png)

`PR_curve`
![image](https://user-images.githubusercontent.com/8830319/162952853-7b52113d-03b6-4626-b01d-ec294ac624ee.png)

`F1_curve`
![image](https://user-images.githubusercontent.com/8830319/162952890-46543679-46f1-432b-a7f3-d20ec68239d2.png)

`Confusion Matrix`
![image](https://user-images.githubusercontent.com/8830319/162952936-a889e739-3c6a-4000-921b-4ebdebc613a1.png)

**Comment by glenn-jocher** [comment]
@sarmientoj24 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Verify labels.** View `train_batch*.jpg` to verify that your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by sarmientoj24** [comment]
```
Verify labels. View train_batch*.jpg to verify that your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
```

I am actually not getting `train_batch*.jpg`

**Comment by glenn-jocher** [comment]
@sarmientoj24 train_batch*.jpg are created in your runs/exp/train directory on training start. First 3 batches are saved after they run.

**Comment by fatejzz** [comment]
Sorry to bother you, have you ever solved this problem since I also have this question.

**Comment by ruman1609** [comment]
Try do geometrical augmentation (rotate 90°, flipping, etc.).

After that when you training your data, please set shuffle parameter in dataloader to True.

in my case it really help, my validation mAP become 72% right now.

**Comment by glenn-jocher** [comment]
@ruman1609 yes, applying geometric augmentations such as rotation, flipping, and other transformations can help improve the training performance of YOLOv5. These augmentations can increase the variety and diversity of the training data, leading to better generalization capabilities.

Additionally, setting the `shuffle` parameter in the dataloader to True can also contribute to better results. Shuffling the training data ensures that the model is exposed to a diverse range of samples throughout the training process, preventing any bias or pattern that may exist in the data from affecting the model's performance.

By incorporating these techniques, you were able to achieve an impressive validation mAP of 72%. Keep in mind that the effectiveness of these strategies can vary depending on the specific dataset and problem you are working on, so it's always recommended to experiment and fine-tune these parameters based on your specific requirements.

Best regards
